docs/design.md

```diff
@@ -0,0 +1,144 @@
+- The Kubernetes Edge service passes data from Producers (e.g. Firefox) to a
+  PubSub Raw Topic.
+   - The service will store data on disk when PubSub is unavailable.
+      - New messages always attempt delivery to PubSub first.
+      - Data on disk is retried periodically.
+      - The service cannot be autoscaled down when there is data on disk.
+   - The service will respond 500 if PubSub and disk space are both unavailable.
+      - This is to be avoided because it has numerous negative side-effects,
+        including delayed `submission_date` and usually some amount of data
+loss.
+   - The PubSub Topic for data is configurable based on `uri`.
+   - Expected topics are Structured Ingestion, Telemetry, and Pioneer.
+- The Dataflow Landfill Sink job copies data from a PubSub Raw Topic to Cloud
+  Storage.
+   - Data can be read from Cloud Storage for backfill, recovery, or testing.
+- The Dataflow Decoder job decodes data from a PubSub Raw Topic to a PubSub
+  Decoded Topic.
+   - Decoding data applies these transforms in order:
+      - Parse `uri` attribute into multiple attributes.
+      - Gzip decompress `payload` if gzip compressed.
+      - Validate `payload` using a JSON Schema determined by metadata fields.
+      - Resolve GeoIP from `remote_addr` or `x_forwarded_for` attribute into
+        `geo_*` attributes.
+      - Parse `agent` attribute into `user_agent_*` attributes.
+      - Copy attributes to `metadata` top level key in `payload`.
+      - Remove duplicates based on `document_id` using Cloud Memorystore
+        (Redis).
+   - The job sends invalid data, including duplicates, to a configurable error
+     output.
+      - The error output can be a Cloud Storage path or a PubSub Topic.
+- The Dataflow BigQuery Sink job copies data from a PubSub Decoded Topic to
+  BigQuery.
+   - The job also sends rejected data to a configurable error output.
+      - The error output can be a Cloud Storage path or a PubSub Topic.
+- The Dataflow Dataset Sink job copies data from a PubSub Decoded Topic to Cloud
+  Storage.
+   - BigQuery can load data to backfill columns previously unspecified in the
+     schema.
+   - BigQuery, Spark, and Dataflow can load data to execute ad-hoc queries.
+- The Dataflow `DocType` Splitter job splits data from a PubSub Decoded Topic to
+  multiple PubSub `DocType` Topics.
+   - This job is only configured for requested `DocType`s.
+
+### Notes
+
+PubSub stores unacknowledged messages for 7 days. Any PubSub subscription more
+than 7 days behind requires a backfill.
+
+Dataflow will extend ack deadlines indefinitely when consuming messages, and
+will not ack messages until they are processed by an output or GroupByKey
+transform.
+
+Dataflow jobs achieve at least once delivery by *not* using GroupByKey
+transforms and *not* falling more than 7 days behind in processing.
+
+## Design Decisions
+
+### Kubernetes and PubSub
+
+They are simple, scalable, managed services. Using them as opposed to a compute
```

Comment:
I hesitate to describe Kubernetes as "simple". It presents a relatively simple and now industry-standard interface for complex orchestration.

Perhaps "they are scalable managed services that are well-integrated standards in the GCP environment

Reply:
I updated the wording here to describe PubSub and Kubernetes separately